### PR TITLE
On-Chain Config App initializer[sc-3242]

### PR DESCRIPTION
**Pull request by deep-quality-dev**
### description of pull request:

Create `morph` app that works as oracle-suite initializer, which is responsible for 
- keeping up-to-dated config from `ConfigRegistry`
- running app with up-to-dated config.
- if on-chain config has been updated, restart app via sending interrupt signal.

#### Run `morph` with embedded config(`config-morph.hcl`)
```bash
./morph run
```

#### Run `morph` with arguments
```bash
./morph run --bin ./ghost --config config-cache.hcl --config-rpc [rpc url] --config-registry [ConfigRegistry address]
```
With this command, morph will prioritize the arguments to run.
`--bin`: required, indicates a command to run application
`--config`: optional, indicates a path to the local config file being loaded when app starts running
`--config-rpc`: optional, indicates a rpc url to be used to fetch on-chain config from ConfigRegistry contract
`--config-registry`: optional, indicates a address to ConfigRegistry contract

### CODE REVIEW

File: pkg/morph/morph.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+		m.log.WithError(err).Error("Failed writing in-memory config to the file:" + m.morphFile)
+		return false, err
+	}
+
+	m.log.Info("Found that on-chain configuration has been updated")
+
+	return true, nil
+}
+
+func (m *Morph) RestartApp() error {
```
Comment (teghnet) [inline pkg/morph/morph.go:218]: I think we discussed, that we want the app to exit and let k8s (or whatever orchestration we have) restart it.

**Review by teghnet — CHANGES_REQUESTED** [review]
Generally, we need more documentation on how this code is to behave.
Not all is obvious.

File: pkg/supervisor/appmanager.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+		arguments:              cfg.Arguments,
+		waitDurationForQuiting: cfg.WaitDurationForQuiting,
+		log:                    cfg.Logger.WithField("tag", AppManagerLoggerTag),
+	}, nil
+}
+
+// Start checks if application was already running and make sure that app is running
+// If running, take its process, else execute command to run application
+func (am *AppManager) Start(ctx context.Context) error {
+	am.ctx = ctx
```
Comment (teghnet) [inline pkg/supervisor/appmanager.go:93]: Let's check if app is already running.
ie. check if ctx is not null

Comment (deep-quality-dev) [reply pkg/supervisor/appmanager.go:93]: I need to check if app is already running. As `morph` initiate the `ghost` at the beginning, if `morph` was killed or crashed for some reason, `morph` might rerun app at the next running. 
So in order to keep only one process is running, I need to kill the existing process before start running `ghost`.

File: pkg/morph/morph.go
[the remark this answers is not in the document]
Comment (deep-quality-dev) [reply pkg/morph/morph.go:1]: Replaced `morph` in pkg with `watchdog`

File: cmd/morph/cmd_run.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			if len(options.rpcURL) > 0 {
+				os.Setenv("CFG_CHAIN_RPC_URLS", options.rpcURL)
+			}
+			if len(options.bin) > 0 {
+				os.Setenv("CFG_APP_BIN", options.bin)
+				os.Setenv("CFG_APP_ARGS", options.args)
+			}
+			if len(options.configRegistry) > 0 {
+				os.Setenv("CFG_CONFIG_REGISTRY", options.configRegistry)
+			}
```
[the remark this answers is not in the document]
Comment (deep-quality-dev) [reply cmd/morph/cmd_run.go:60]: Removed setting env vars and we can use same `run` command with other apps.

File: pkg/supervisor/appmanager.go (hunk 2 above)
Comment (teghnet) [reply pkg/supervisor/appmanager.go:101]: I mean 
```go
	if s.ctx != nil {
		return errors.New("service can be started only once")
	}
	if ctx == nil {
		return errors.New("context must not be nil")
	}
```

**Pull request closed without merging**